api/arrange.py

```diff
@@ -1,34 +1,42 @@
 import bpy
 import typing
+from collections import deque
 
 def _arrange(node_tree, padding: typing.Tuple[float, float] = (50, 25)):
     # Organize the nodes into columns based on their links.
     columns: typing.List[typing.List[typing.Any]] = []
-    def contains_link(node, column):
-        return any(
-            any(
-                any(link.from_node == node for link in input.links)
-                for input in n.inputs
-            )
-            for n in column
-        )
-    for node in reversed(node_tree.nodes):
-        if (x := next(
-            filter(
-                lambda x: contains_link(node, x[1]),
-                enumerate(columns)
-            ),
-            None
-        )) is not None:
-            if x[0] > 0:
-                columns[x[0] - 1].append(node)
-            else:
-                columns.insert(x[0], [node])
+    
+    def topo_sort(graph):
+        in_degree = {u: 0 for u in graph}
+        for u in graph:
+            for v in graph[u]:
+                in_degree[v] += 1
+        queue = deque([u for u in in_degree if in_degree[u] == 0])
+        topo_order = []
+        while queue:
+            u = queue.popleft()
+            topo_order.append(u)
+            for v in graph[u]:
+                in_degree[v] -= 1
+                if in_degree[v] == 0:
+                    queue.append(v)
+        return topo_order
+    
+    graph = { node:set() for node in node_tree.nodes }
+    node_input_link_count = { i:0  for node in node_tree.nodes for i in node.inputs }
```

Comment:
Might be able to use `collections.Counter` here.

Reply:
Ah yeah that's clean...